### PR TITLE
CODEOWNERS trailing slash matching semantics

### PR DESCRIPTION
**Pull request by cbart**
Semantics of CODEOWNERS file matching for a rule with a trailing slash:

- `/foo/bar/` should match like `/foo/bar/**` - that is any file with the prefix on its path is matched
- However neither `/foo/bar/` nor `/foo/bar/**` match the file `/foo/bar` as the trailing slash indicates `bar` needs to be treated as a directory.

Here is a bit of example from github documentation which motivates this semantics:

```
# In this example, @doctocat owns any files in the build/logs
# directory at the root of the repository and any of its
# subdirectories.
/build/logs/ @doctocat
```

There is another example (which I think is incorrect misleading) where a negative pattern not ending with trailing slash is suggested to match the whole sub-directory hierarchy as if it had a trailing `/`:

```
# In this example, @octocat owns any file in the `/apps`
# directory in the root of your repository except for the `/apps/github`
# subdirectory, as its owners are left empty.
/apps/ @octocat
/apps/github
```

## Test plan

Expand the unit tests with relevant examples.

### CODE REVIEW

File: internal/own/codeowners/proto/find_owners.go [hunk 1]
```diff
@@ -78,14 +78,19 @@ func compile(pattern string) (globPattern, error) {
...
-	// However, this pattern does not match if the path finishes
-	// with the part that matches token preceeding `/`. This is why
-	// the extra `*` is needed, and `/**` will not suffice.
+	// Trailing `/` is equivalent with ending the pattern with `/**` instead.
 	if strings.HasSuffix(pattern, separator) {
-		glob = append(glob, anySubPath{}, anyMatch{})
+		glob = append(glob, anySubPath{})
+	}
+	// Trailing `/**` (explicitcly or implicitly like above) is neccessarily
+	// transleted to `/**/*.
```
Comment (eseliger) [inline internal/own/codeowners/proto/find_owners.go:86]:
```suggestion
	// translated to `/**/*.
```

File: internal/own/codeowners/proto/find_owners.go [hunk 2]
```diff
@@ -78,14 +78,19 @@ func compile(pattern string) (globPattern, error) {
...
-	// anchored at the directory that the pattern describes.
-	// However, this pattern does not match if the path finishes
-	// with the part that matches token preceeding `/`. This is why
-	// the extra `*` is needed, and `/**` will not suffice.
+	// Trailing `/` is equivalent with ending the pattern with `/**` instead.
 	if strings.HasSuffix(pattern, separator) {
-		glob = append(glob, anySubPath{}, anyMatch{})
+		glob = append(glob, anySubPath{})
+	}
+	// Trailing `/**` (explicitcly or implicitly like above) is neccessarily
```
Comment (eseliger) [inline internal/own/codeowners/proto/find_owners.go:85]:
```suggestion
	// Trailing `/**` (explicitly or implicitly like above) is neccessarily
```

**Review by eseliger — APPROVED** [review]
Good stuff!

File: internal/own/codeowners/proto/find_owners_test.go [hunk 3]
```diff
@@ -94,6 +94,19 @@ func TestFileOwnersNoMatch(t *testing.T) {
...
+				// These do not match as the right-most directory name `leaf`
+				// is just a prefix to the corresponding directory on the given path.
+				"/directory/leaf_and_more/file",
+				"/prefix/directory/leaf_and_more/file",
+				// These do not match as the pattern matches anything within
+				// the sub-directory tree, but not the directory itself.
+				"/directory/leaf",
+				"/prefix/directory/leaf",
+			},
+		},
```
Comment (leonore) [inline internal/own/codeowners/proto/find_owners_test.go:109]: from what I understand `directory/leaf/` and `directory/leaf/**` will function the same, can you add a unit test for the matching case just for symmetry and confirmation? (confused me a little)
```go
		{
			pattern: "directory/leaf/**",
			paths: []string{
				"/directory/leaf/",
                 "/directory/leaf/sub_leaf",
			},
		},
```

**Review by leonore — APPROVED** [review]
thinking about regexp for too long is headache inducing but lgtm!

File: internal/own/codeowners/proto/find_owners_test.go (hunk 3 above)
Comment (cbart) [reply internal/own/codeowners/proto/find_owners_test.go:118]: Good catch!

**Pull request merged**